larrecodnn/ImagePatternAlgs/Modules/RawWaveformClnSigDump_module.cc

```diff
@@ -0,0 +1,815 @@
+  {
+    return static_cast<bool>(generator_names.count(gname));
+  };
+  std::string
+  get_gen(int tid)
+  {
+    if (!isSorted) { this->sort_now(); }
+    return std::lower_bound(track_id_map.begin(),
+                            track_id_map.end(),
+                            tid,
+                            [](const auto& a, const auto& b) { return (a.first < b); })
+      ->second;
+  };
+};
+
+// Create the random number generator
+namespace {
+  std::string const instanceName = "RawWaveformClnSigDump";
+}
+
+//-----------------------------------------------------------------------
+nnet::RawWaveformClnSigDump::RawWaveformClnSigDump(fhicl::ParameterSet const& p)
+  : EDAnalyzer{p}
+  , fDumpWaveformsFileName(p.get<std::string>("DumpWaveformsFileName", "dumpwaveforms"))
+  , fDumpCleanSignalFileName(p.get<std::string>("CleanSignalFileName", "dumpcleansignal"))
+  , fSimulationProducerLabel(p.get<std::string>("SimulationProducerLabel", "larg4Main"))
+  , fSimChannelLabel(p.get<std::string>("SimChannelLabel", "elecDrift"))
+  , fDigitModuleLabel(p.get<std::string>("DigitModuleLabel", "simWire"))
+  , fCleanSignalDigitModuleLabel(p.get<std::string>("CleanSignalDigitModuleLabel", "simWire:signal"))
+  , fUseFullWaveform(p.get<bool>("UseFullWaveform", true))
+  , fShortWaveformSize(p.get<unsigned int>("ShortWaveformSize"))
+  , fEstIndFWForOffset(p.get<int>("EstIndFWForOffset"))
+  , fEstColFWForOffset(p.get<int>("EstIndFWForOffset"))
+  , fSelectGenLabel(p.get<std::string>("SelectGenLabel", "ANY"))
+  , fSelectProcID(p.get<std::string>("SelectProcID", "ANY"))
+  , fSelectPDGCode(p.get<int>("SelectPDGCode", 0))
+  , fPlaneToDump(p.get<std::string>("PlaneToDump"))
+  , fMinParticleEnergyGeV(p.get<double>("MinParticleEnergyGeV", 0.))
+  , fMinEnergyDepositedMeV(p.get<double>("MinEnergyDepositedMeV", 0.))
+  , fMinNumberOfElectrons(p.get<int>("MinNumberOfElectrons", 600))
+  , fMaxNumberOfElectrons(p.get<int>("MaxNumberOfElectrons", -1))
+  , fMinPureSignalADCs(p.get<int>("MinPureSignalADCs", 0))
+  , fSaveSignal(p.get<bool>("SaveSignal", true))
+  , fMaxNoiseChannelsPerEvent(p.get<int>("MaxNoiseChannelsPerEvent"))
+  , fCollectionPlaneLabel(p.get<std::string>("CollectionPlaneLabel"))
+{
+  if (std::getenv("CLUSTER") && std::getenv("PROCESS")) {
+    fDumpWaveformsFileName += string(std::getenv("CLUSTER")) + "-" + string(std::getenv("PROCESS")) + "-";
+    fDumpCleanSignalFileName += string(std::getenv("CLUSTER")) + "-" + string(std::getenv("PROCESS")) + "-";
+  }
+
+  if (fDigitModuleLabel.empty() && fCleanSignalDigitModuleLabel.empty()) {
+    throw cet::exception("RawWaveformClnSigDump")
+      << "Both DigitModuleLabel and CleanSignalModuleLabel are empty";
+  }
+  fRandFlat = std::make_unique< CLHEP::RandFlat >
+              (createEngine(art::ServiceHandle<rndm::NuRandomService>{}->declareEngine(
+                            instanceName, p, "SeedForRawWaveformDump"),
+                            "HepJamesRandom", instanceName));
```

Comment:
I suggest a simpler approach, Mike.  I'd change the type of `fRandFlat` to `CLHEP::RandFlat` and initialize it in the initialization list of the class:

```cpp
, fRandFlat{createEngine(art::ServiceHandle<rndm::NuRandomService>{}->declareEngine(
                         instanceName, p, "SeedForRawWaveformDump"),
                         "HepJamesRandom", instanceName)}
```

The `declareEngine` bit is ugly, but I don't have an easy solution for you right now.

Reply:
thanks Kyle for the recommendations, I just implemented them in the updated pr.